src/plugins/validation.ts

```diff
@@ -0,0 +1,388 @@
+import * as os from 'os';
+import { table } from 'table';
+import { ValidationConfig } from '../config';
+import { PluginManager } from './_manager';
+
+/**
+ * Context available to plugins during validation.
+ */
+export class ValidationContext {
+
+  private readonly _report: ValidationReport;
+  private readonly _logger: ValidationLogger;
+
+  constructor(
+
+    /**
+     * The list of manifests to validate.
+     */
+    public readonly manifests: string[],
+
+    /**
+     * The NodeJS package name of the plugin running the validation.
+     */
+    public readonly pkg: string,
+
+    /**
+     * The version of the NodeJS package that runs the validation.
+     */
+    public readonly version: string,
+
+    /**
+     * Construct metadata of resources in the application.
+     *
+     * @default - No metadata. This means construct aware metadata will not be available in the report.
+     */
+    public readonly metadata?: {[key: string]: ResourceConstructMetadata},
+
+    /**
+     * Whether or not the synth command was executed with --stdout.
+     *
+     * @default false
+     */
+    public readonly stdout?: boolean) {
+
+    this._report = new ValidationReport(this.pkg, this.version, this.metadata ?? {}, stdout ?? false);
+    this._logger = new ValidationLogger();
+  }
+
+  /**
+   * Report emitted by the validation.
+   *
+   * Plugins should interact with this object to generate the report.
+   */
+  public get report(): ValidationReport {
+    return this._report;
+  }
+
+  /**
+   * Logger for the validation.
+   *
+   * Plugins should interact with this object to log messages during validation.
+   */
+  public get logger(): ValidationLogger {
+    return this._logger;
+  }
+}
+
+/**
+ * Logger available to plugins during validation. Use this instead of `console.log`.
+ */
+export class ValidationLogger {
+
+  /**
+   * Log a message.
+   */
+  public log(message: string) {
+    console.log(message);
+  }
+}
```

Comment:
So like we talked about, I initially did this so that we can control `synth` output for the `--json`. I decided eventually to allow specifying a file where the report will be outputed to in JSON format. 

Doing this via `stdout` was just too big of a change at this time, and we can always add it later if needed. The file is simple and straight forward.